### ISSUE TITLE
THC_createRandomSeed is not thread-safe

### ISSUE THREAD

**Issue by ebetica** [pytorch/pytorch#6288]
When we call an ATen `type.randn({n});` function from multiple threads simultaneously, there's a chance it errors with 
```
THCudaCheck FAIL file=/private/home/kaia/FAIR_rush/autogradpp/pytorch/aten/src/THC/generic/THCTensorMathPairwise.cu line=81 error=77 : an illegal memory access was encountered terminate called after throwing an instance of 'std::runtime_error' what(): Creating MTGP kernel state failed. at /private/home/kaia/FAIR_rush/autogradpp/pytorch/aten/src/THC/THCTensorRandom.cu:38
```
This problem disappears if you try generate a random number on each device before you launch any threads.

I suspect it's because [this](https://github.com/torch/cutorch/blob/e2051b652d5b1f5182a1bfce11da9f53c2f92bd8/lib/THC/THCTensorRandom.cpp#L79) function is not protected by a mutex:

**Comment by ezyang** [comment]
Someone else tried to report this to us at #4777 but it seems likely we accidentally closed this as a dupe of CPU RNG not being thread safe.

CPU RNG used to not be thread safe but https://github.com/pytorch/pytorch/pull/4041 made it thread safe.

**Comment by ebetica** [comment]
I made a reproducible test, it seems like this only happens when you try to generate random numbers on two different GPUs with two different threads at the same time.

https://github.com/ebetica/autogradpp/blob/buggy_pytorch/tests/misc_t.cpp#L39-L53

**Comment by yf225** [comment]
Fixed in #6527
